Add getFileDestinationUrl endpoint

slifty opened the pull request:
## Description
This PR adds a `POST /api/getFileDestinationUrl` endpoint to the service, which can be used to generate a signed S3 request.

Resolves #7 

## Steps to Test
- Install new dependencies (`npm i`)
- Configure AWS credentials with write access to an S3 bucket.
- POST to the new endpoint; the post body must include a `bucket` and a `fileType` parameter.
- The API should return information that can then be used to POST your uploaded file directly to S3.

You can also run the integration tests:

- `npm test`

jasonaowen commented on src/controllers/fileDestinationUrl.controller.ts:
```diff
@@ -0,0 +1,25 @@
+): void => {
+  fileDestinationUrlService.createFileDestinationUrl(req.body)
+    .then((data) => res.json(data))
+    .catch((err) => res.status(500).json(err));
```
This isn't capturing errors as I'd expect; I'm just getting empty objects in the response.

...[what an unexpectedly complicated situation](https://stackoverflow.com/questions/18391212/is-it-not-possible-to-stringify-an-error-using-json-stringify)!

Seems like [serialize-error](https://www.npmjs.com/package/serialize-error) might be the simplest solution? Assuming we do in fact want to publish our uncaught errors.

slifty replied:
Oh interesting!  I swore it was posting errors but maybe that was an earlier draft of the code (oh hey tests!)

Will dig in.

slifty replied:
Rather than import a dependency I just made a little utility method that will serialize the error.  This has the added benefit of stripping the stack since I don't think we want to expose that to API clients anyway.

jasonaowen reviewed (changes requested):
This looks like a good start!

There are two things I'd love to see included: tests, and health.

Tests imply some kind of dependency injection, since we're interacting with an external service. I haven't thought too deeply about this yet, but we can at least test the shape of the required parameters discussed below.

More importantly, now that there's some required configuration - AWS credentials - the health endpoint should verify that it is configured correctly. Maybe the lowest lift of that would be simply that it is configured at all; better would be that the credentials are valid and have the necessary permissions. Unfortunately, I don't know the SDK well enough to say how to do that or even if it's possible, so I guess this is a nice-to-have, but if it is possible now might be a good time to add them!

jasonaowen commented:
> This PR is missing:
> 
>     1. Documentation
> 
>     2. Tests
> 
> 
> But I believe it is ready for feedback in terms of the functionality.

Oh derp, you specifically mentioned that in the description! And then I went and asked for both in my review. :stuck_out_tongue_closed_eyes: Sorry about that!

slifty commented on src/services/fileDestinationUrl.service.ts:
```diff
@@ -0,0 +1,49 @@
+}: {
+  bucket: string;
+  fileType: string;
+  maxSize?: number;
+  fileName?: string;
+  path?: string;
```
Ah -- when we spoke on that I was scoped to just the single parameter under discussion (fileType).  I didn't realize we were talking about NO optional parameters.

I'm OK with `maxSize` being required for sure.

I believe that `path` makes more sense as optional with a default of `/` -- I'm not sure what we gain from requiring `/` to be passed.

`fileName` being required feels like the incorrect API design (file names can have collisions in a bucket).  Allowing fileName allows a client to explicitly write over an existing file, but requiring fileName makes it possible for a client to *accidentally* write over an existing file.

NOTE: Since I moved to POST requests aws might not actually allow data to overwrite (gonna test that now), but even still this would make the calling client responsible for avoiding collisions despite having no way of discovering what is / is not a collision.

jasonaowen replied:
Ah sorry, there are two parts here which I conflated a bit!

1. Should the TypeScript type definition for this method allow for optional parameters? This is what I meant to focus on with this review comment! Since we're providing default values in the method, the *type* of these arguments cannot ever be `undefined`: they're either the value the caller passed in, or the default value provided. In this narrow technical context, I think it's safe and uncontroversial to say that these should all be non-optional.
1. Should the API require these parameters? This is the broader question, and I'm glad you noted it, since I didn't realize we weren't completely on the same page! I'm generally of the opinion that fewer parameters - and especially, fewer optional parameters - is better. I think we could drop `fileName` and only use UUIDs; particularly if upon further research you discover that you cannot overwrite via POST. Dropping `path` is less clear, particularly since we need to support some kind of path prefix for the S3 bucket for the local dev environments; that could be optional, although I'm also slightly tempted to argue for an environment variable. I haven't yet thought through what the local development process is going to look like for integrating this service.

slifty replied:
The environment variable question is a great one to raise; I actually was on the fence about it for bucket being defined by environment as well.

I'll finish the first pass updating based on this feedback / thought process and we can continue the discussion from there.

slifty replied:
for future reference: POST *does* allow you to overwrite files.

slifty commented on src/services/fileDestinationUrl.service.ts:
```diff
@@ -0,0 +1,49 @@
+const createFileDestinationUrl = async ({
+  bucket,
+  fileType,
+  maxSize = 0,
```
Good catch -- meant to add a conditional to that; moot point since we're going to require maxSize anyway.

slifty commented on src/controllers/fileDestinationUrl.controller.ts:
```diff
@@ -0,0 +1,25 @@
+  };
+}
+
+const createFileDestinationUrl = (
+  req: CreateFileDestinationRequest,
+  res: Response,
+): void => {
+  fileDestinationUrlService.createFileDestinationUrl(req.body)
+    .then((data) => res.json(data))
+    .catch((err) => res.status(500).json(err));
+};
```
Great point -- so there are two paths for validation.

The first is to ignore TypeScript and just use something like [express-validator]()https://express-validator.github.io/docs/.  The benefit of this approach is that we have more control over the validation rules beyond just data type (e.g. we can require positive integers, email addresses, etc).

The second is to use an [approach like this one](https://www.worl.co/2016/12/building-an-api-backend-with-typescript-and-express---part-two-validation/) which basically creates a custom validator and uses a library that maps json to TypeScript.

I'm inclined to use the `express-validator` approach, which feels more explicit and given the point you raise that TypeScript is not actually meant to provide runtime validation I don't think it is a violation of DRY either.

jasonaowen replied:
The custom validator post starts out with "enable this experimental feature", and that feature, four years later, appears to [still be experimental in TypeScript](https://www.typescriptlang.org/docs/handbook/decorators.html) and is a [Stage 2 proposal in JavaScript](https://github.com/tc39/proposal-decorators). I've worked with codebases that use experimental JavaScript features in the past, and found that the use tends to end up being more work than the lines of code saved, so let's not do that.

(The custom validator approach blog post wasn't loading for me; here's the wayback link for convenience: [Building an API Backend with TypeScript and Express - Part Two: Validation](https://web.archive.org/web/20191013032705/https://www.worl.co/2016/12/building-an-api-backend-with-typescript-and-express---part-two-validation/).)

I think it's worth clarifying that I'm not advocating that we ignore TypeScript; quite the opposite. Basically, any time we override TypeScript's type inference is a red flag; we're telling it we know more than it does about the types we're working with, and we need to be really really careful about that. Here, that took the form of `req: CreateFileDestinationRequest`. I don't know Express super well, and this is a subtle boundary between framework code and our code, but my initial instinct is to use the provided types (`req: Request`), validate them, and only then cast to our expected type.

Joi looks awesome, we should definitely use that - nice find! Should make the validation step a lot easier to write and easier to read. I haven't quite wrapped my head around express-validator yet, particularly in how it works with types; it definitely feels worth exploring! I'd be open to hearing either that it's way better to use it, or that it turns out not to be much better than doing the validation ourselves in the method.

slifty replied:
I've started implementing validation using Joi (experess-validator would be an alternative to Joi; it's way less popular and way more limited, so I think Joi is the right path)

slifty commented on src/services/fileDestinationUrl.service.ts:
```diff
@@ -0,0 +1,49 @@
+    ],
+  };
+
+  const presignedPost: S3.PresignedPost = await new Promise<S3.PresignedPost>((resolve, reject) => {
+    s3.createPresignedPost(s3Params, (err: Error|null, data) => {
+      if (err) reject(err);
+      resolve(data);
+    });
+  });
```
Great suggestion!  Moving to a helper method now.

Also just because I did the research originally: AWS SDK actually *does* have a promise based API... but... it is inconsistent.

1. Several AWS sdk methods let you tack on a [`.promise()`](https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/using-promises.html) at the end of any method that generates an AWS request, which will convert the call to a promise. 

2. getSignedUrl (which isn't what we are using, since that is for PUT not POST) does not support `.promise` but there is a [`.getSignedUrlPromise()`](https://github.com/aws/aws-sdk-js/pull/2827) which does.

3. `createPresignedPost` does not return an AWS request, so it doesn't have the `.promise()` option, and nobody has added a `getSignedUrlPromise` equivalent (that I could find).

Long story short, ☠️

jasonaowen replied:
Thanks, and nice research! :skull_and_crossbones: indeed.

slifty commented on src/services/fileDestinationUrl.service.ts:
```diff
@@ -0,0 +1,49 @@
+  };
+
+  const presignedPost: S3.PresignedPost = await new Promise<S3.PresignedPost>((resolve, reject) => {
+    s3.createPresignedPost(s3Params, (err: Error|null, data) => {
```
I found [a bug with the aws sdk](https://github.com/aws/aws-sdk-js/issues/3486) which can only be avoided if I change the way that I call `createPresignedPost` to not use a callback.

However... the sdk notes that:

```
   * @note You must ensure that you have static or previously resolved
   *   credentials if you call this method synchronously (with no callback),
   *   otherwise it may not properly sign the request. If you cannot guarantee
   *   this (you are using an asynchronous credential provider, i.e., EC2
   *   IAM roles), you should always call this method with an asynchronous
   *   callback.
```

Ultimately we have a choice:

1. Use async, which allows us to leverage EC2 IAM roles, but will cause an un-catchable fatal error if the service is deployed in an environment without apppropriate credentials.

2. Use sync, which will limit how we can set AWS credentials for this service.

slifty commented:
At long last this is ready for a re-review.  I have now added:

1. Runtime validation
2. Integration tests
3. Updated health check
4. Other requested changes

I started to add, but decided to punt on for this minute, API documentation using OpenAPI spec.  I think this belongs in the project (and normally would belong in the PR that defines a new endpoint), but since it involves a little more lift around pattern-setting I would like to make it a separate PR.

xmunoz commented on README.md:
````diff
@@ -18,7 +22,29 @@ This service does not process files, but provides the information necessary to t
+}
+```
+
+3. Start the project.
````
Can you include a brief code-snippet here of how to actually use this endpoint?

xmunoz commented:
> I started to add, but decided to punt on for this minute, API documentation using OpenAPI spec

I think this is overkill. Just a code snippet in the README should suffice.

jasonaowen reviewed (approved):
The code looks good to me! Thank you so much, @slifty, this is excellent.

I was able to successfully test the health API; I added some `console.log` debugging output to make sure it was working as expected, which was very helpful.

I was also able to test the parameter validation. Joi is really great! Wonderful find; very straightforward code on our part, and very clear output when validation fails. Example command:

```sh
$ curl -v -X POST --data '{"bucket":null, "fileType": true, "maxSize": "five"}' -H "Content-Type: application/json" http://localhost:3000/api/fileDestinationUrl | jq .
```

Finally, I was able to partially test the S3 signed request generation. It did *something* when I gave it correct parameters, and the response looks like what I'd expect. Unfortunately, I'm not entirely sure what to *do* with the response, and I'm pretty sure that's a problem with my lack of knowledge rather than with the code itself.

I tried constructing a curl(1) request like so:

```sh
$ curl -vvv -L -X POST \
  -F 'Key=...' -F 'bucket=...' -F 'X-Amz-Algorithm=...' \
  -F 'X-Amz-Credential=...' -F 'X-Amz-Date=...' -F 'Policy=' \
  -F 'X-Amz-Signature=...' -F 'file=@test.png \
  $(URL)
```

and manually substituting the various values from the API response. When using the URL returned at `.destinationUrl`, I got a 405 Method Not Allowed. When using the URL returned at `.presignedPost.url`, I got a 301 Moved Permanently with the message "The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint." When using the URL from the redirect message, I got 400 Bad Request with a message "the region 'us-east-1' is wrong; expecting 'us-west-2'" - presumably referring to the value of `.presignedPost.fields.X-Amz-Credential`, which does say `us-east-1` in it.

However, I believe that you were able to use the response successfully, @slifty, and the code appears to be working, so I'm comfortable voting +1 to merge. Let's talk later about what I should have been doing to test this!

slifty commented on src/validators/validateCreateFileDestinationUrlParams.ts:
```diff
@@ -0,0 +1,15 @@
+  data: CreateFileDestinationUrlParams,
+): Joi.ValidationResult => Joi.object().keys({
+  bucket: Joi.string().min(3).max(63).required(),
+  fileName: Joi.string().max(1024),
```
Oh great catch Jason!  Adding that fix before merging.
